Plans: highlight the current plan in the plan features grid

aneeshd16 opened the pull request:
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Fixes Automattic/martech#1636

## Proposed Changes

* Adds a new card above the pricing grid that highlights the current plan on the /plans page.
* The card is also shown if the current plan is a monthly plan and the user selects "Pay Annually" via the toggle. In this case, the card shows the annual plan details.

Figma: dyXNnR0oHu8H47eUES8ktJ-fi-385%3A39202

## Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->


* On a site with a free plan, go to `/plans/<site slug>`. Confirm that the plan card is shown and matches the design.
<img width="1374" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/83800175-0a86-41b1-b6ce-f8235be53aee">

-----

* On a site with a monthly paid plan, go to `/plans/<site slug>`. Confirm that the plan card is shown and matches the design.
<img width="1341" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/a133f21d-fa63-41e0-ad60-e4e735d77ea1">

* Click on the "Pay annually" option in the toggle. Confirm that the plan card is shown and has the details for the yearly equivalent of the current monthly plan.
<img width="1341" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/fc6e8293-65f7-41c2-b893-c3067207fc73">

-----

* On a site with a 1Y/2Y/3Y paid plan, go to `/plans/<site slug>`. Confirm that the plan card is shown and matches the design.
<img width="1418" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/c15f6a47-57d4-4f71-9aed-dc30ad5ef321">

### Testing for regressions
Check the following flows and confirm that the current plan card is NOT shown.
* `/plans/yearly/<site slug>?domain=true&domainAndPlanPackage=true`
* `/start/plans`
* `/setup/newsletter/plans`
* `/start/launch-site?siteSlug=<site slug>`

## Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [ ] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [ ] Have you checked for TypeScript, React or other console errors?
- [ ] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?

chriskmnds reviewed (dismissed):
Can we rebase this onto https://github.com/Automattic/wp-calypso/pull/78731 as it's doing a bunch of cleanup of redundant props with this component. it should be respected as the cleaner/latest form where there's no breakage. so I'd appreciate it if we can prioritize https://github.com/Automattic/wp-calypso/pull/78731 over any additional restructuring/implementations 🙏 

(and hopefully there shouldn't be any conflicts - just seeing some cleanup that's going into the plans properties computation and not sure if it will lead to unnecessary conflicts on other end)

cc @Automattic/martech-decepticons

aneeshd16 commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -913,161 +956,179 @@ const ConnectedPlanFeatures2023Grid = connect(
 		const selectedSiteSlug = getSiteSlug( state, siteId );
+		let currentSitePlanProperties = null;
+
+		const planProperties: PlanProperties[] = plans.reduce(
```
This entire change is essentially changing the `.map` to `.reduce`.

An alternative to this was two loops:
```js
filteredPlanProperties = planProperties.filter( plan => !( isCurrentPlan || getPlanClass( plan ) === getPlanClass( currentSitePlanSlug ) );
currentSitePlanProperties = planProperties.find( plan => ( isCurrentPlan || getPlanClass( plan ) === getPlanClass( currentSitePlanSlug ) );
```

If we agree that the above implementation is easier to read, I'll be happy to change this.

chriskmnds replied:
Cool. My concern is it might create a bunch of pointless conflicts in the other PR for a minor change like this. Maybe just do the two loops implementation if it will avoid conflicts. I'd appreciate anything to help prioritize https://github.com/Automattic/wp-calypso/pull/78731. This whole lot will also get refactored soon (migrate out of plans-features-2023-grid).

aneeshd16 replied:
Thanks for highlighting this! If #78731 is ready, I can help review it, and we can merge this one later.

chriskmnds replied:
Thanks @aneeshd16!  https://github.com/Automattic/wp-calypso/pull/78731 is ready. It's mostly tuning types and removing redundant props from typing consuming end.

I'll unblock this PR - If any breakage in https://github.com/Automattic/wp-calypso/pull/78731 then we should proceed as normal with any other work. I'll get back to reviewing your PR tomorrow if none catches me to it. 🙌

ddc22 commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -471,6 +486,34 @@ export class PlanFeatures2023Grid extends Component<
+			</table>
+		);
+	}
+
```
Is there a reason to use a table structure here? I think we can just use generic card like structure with divs. The change seems trivial when using the existing code to bend it towards this requirement. But it seems to feed into the previous pattern of trying to solve everything with a table 😬 .

aneeshd16 replied:
Thanks for the suggestion! I have changed this to a div-based render in https://github.com/Automattic/wp-calypso/pull/78845/commits/b4d2d720779250e28504fe442420cae3734e32c9.

ddc22 replied:
Thanks Aneesh ! This looks better. However, I wonder if the isMobile parameter will confuse us down the line when coming back for a refactor.

Looking back at this code and the related `isMobile` flag. It's basically just deciding whether to render a `td` or `div`
https://github.com/Automattic/wp-calypso/blob/d2e44b92c9b93b95ea6781271c94ec9d51b657a0/client/my-sites/plan-features-2023-grid/index.tsx#L97-L103

Do you think it will be a good idea to probably change this parameter name to `isTableCell` or something similar for now? At least until we come to pick up this refactor? Even though this is temporary, it is unsettling to see a `isMobile` flag hardcoded to true 😬 .

https://github.com/Automattic/wp-calypso/blob/b4d2d720779250e28504fe442420cae3734e32c9/client/my-sites/plan-features-2023-grid/index.tsx#L501

https://github.com/Automattic/wp-calypso/blob/b4d2d720779250e28504fe442420cae3734e32c9/client/my-sites/plan-features-2023-grid/index.tsx#L458

aneeshd16 replied:
Thanks! To keep this PR clean, I have refactored this separately: https://github.com/Automattic/wp-calypso/pull/79605

ddc22 reviewed (commented):
I had a question about the implementation.

chriskmnds commented on client/my-sites/plan-features-2023-grid/style.scss:
```diff
@@ -416,6 +416,45 @@ $mobile-card-max-width: 440px;
+					margin-top: 6px;
+				}
+
+				.plan-features-2023-grid__header-tagline {
+					@include plans-2023-break-medium {
+						min-height: 0;
+						padding-bottom: 16px;
+					}
+				}
+
+				.plan-features-2023-grid__header-logo {
+					@include plans-2023-break-small {
+						margin-top: 18px;
+					}
+				}
```
Is it 100% needed to have these media queries here? I'd love to see us rely less and less on these, even if it means sacrificing a few pixels top/bottom. Most likely these media queries won't be used at all for anything under `plan-features-2023-grid`, probably not in their current form at least (which will require a hefty refactor).

Just asking - can you achieve this without?

aneeshd16 replied:
Good catch! Those weren't needed at all. I have removed them in https://github.com/Automattic/wp-calypso/pull/78845/commits/b4d2d720779250e28504fe442420cae3734e32c9.

chriskmnds commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -471,6 +486,34 @@ export class PlanFeatures2023Grid extends Component<
+	/**
+	 * Similar to `renderTable` above.
+	 */
+	renderCurrentPlan( planPropertiesObj: PlanProperties[] ) {
```
I'd prefer if we rename this to something like `renderPlanHighlight`. 

Although - is it too much work to turn this into a function component instead? That's the type of refactoring that will happen soon in this file (probably by me), so might as well do it here if not too heavy (especially if going with a [non-table view](https://github.com/Automattic/wp-calypso/pull/78845/files#r1252475292))

aneeshd16 replied:
> I'd prefer if we rename this to something like renderPlanHighlight.

Done via https://github.com/Automattic/wp-calypso/pull/78845/commits/d2e44b92c9b93b95ea6781271c94ec9d51b657a0!

aneeshd16 replied:
> Although - is it too much work to turn this into a function component instead? That's the type of refactoring that will happen soon in this file (probably by me), so might as well do it here if not too heavy (especially if going with a [non-table view](https://github.com/Automattic/wp-calypso/pull/78845/files#r1252475292))

I started working on this task by writing a fresh functional component but realized that I was duplicating a lot of existing code and styles. This is a class component, and much of the rendering is done by `renderXYZ` style functions which pick their required props from `this.props`. To reuse these functions, each of these `renderXYZ` functions would have to be converted to functional components. This would be a lengthy task, and I feel it would be best if tackled separately. Definitely, something to do down the line, though!

chriskmnds replied:
Maybe we can do a tiny bit here to help save a few clicks from whoever does that refactor next. Is it possible to create the function component in a separate file and pass these other components through as props or children to render them? 

(or would that lead to more complexity to clean up after?)

aneeshd16 replied:
I believe we should fix the scope and maintain the emphasis on highlighting the current plan in this PR. Introducing refactors may complicate matters for reviewers and testers. Perhaps it would be more convenient to start the refactoring change afresh while looking at the complete picture instead of attempting to integrate a half-baked change into this PR. The new function, "renderPlanHighlight," is small enough to be refactored with ease and utilizes existing functions without introducing any novel concepts.

chriskmnds replied:
> may complicate matters for reviewers and testers

I guess as the reviewer in this case, this would be my suggestion. also in context with everything else happening for [wider refactors](https://github.com/Automattic/wp-calypso/labels/Plans%3A%20NPM%20essential).

aneeshd16 replied:
I'm happy to pick this up in a separate PR, as this particular refactor (class function -> functional component) is not contributing to the goal of the current PR. What do you think about moving this PR forward and tackling this separately?

chriskmnds commented on client/my-sites/plan-features-2023-grid/_media-queries.scss:
```diff
@@ -106,6 +106,13 @@ $planComparisonMinWidthToGridWidthWithSidebarMap: (
+
+	.plan-features-2023-grid__current-plan {
+		display: none;
+		@media ( min-width: $plans-2023-small-breakpoint ) {
+			display: block;
+		}
```
I still can't fathom how we ended up with a design like this (render everything and hide via CSS), and we keep adding to it. This will need to be refactored next.

How big of a deal would it be to move the logic to JS, for this case alone, and pass a variable through from `PlansFeaturesMain` whether to highlight a plan or not (and pass the plan-slug/type to highlight)? Can you think of a minimal solution to just get this case done that way?

aneeshd16 replied:
> I still can't fathom how we ended up with a design like this (render everything and hide via CSS), and we keep adding to it. This will need to be refactored next.

100% agree!

> How big of a deal would it be to move the logic to JS, for this case alone, and pass a variable through from PlansFeaturesMain whether to highlight a plan or not (and pass the plan-slug/type to highlight)? Can you think of a minimal solution to just get this case done that way?

I feel quite strongly about doing this as a separate change which would be easier to review and test instead of combining it with this. If we can add JS logic just for this case, it will be fairly trivial to show/hide the desktop/tablet/mobile views as well. At the top of my mind, this would be a `useGetScreenType` hook which would observe the screen width and return `mobile | tablet | desktop`. After that we could do something like this:
```js
{ screenType === 'desktop' && (
	<div className="plan-features-2023-grid__desktop-view">
		{ this.renderTable( planProperties ) }
	</div>
) }
{ screenType === 'tablet' && (
	<div className="plan-features-2023-grid__tablet-view">
		{ this.renderTabletView() }
	</div>
) }
{ screenType === 'mobile' && (
	<div className="plan-features-2023-grid__mobile-view">
		{ this.renderMobileView( planPropertiesForMobileView ) }
	</div>
) }
```

I have created an issue for tracking this here: https://github.com/Automattic/martech/issues/1912

chriskmnds replied:
I'm not too sure about this. It seems to just let the root problem live on and pass the principle of complexity to the consuming end (to determine which to show when).

So I would wish to see us simplify this to two possible views: `mobile`, and everything else. that's pretty much how it works now, with the "tablet" and "desktop" views being just a difference in number of columns. It's the implementation that suffers, with these over-engineered "screen types", media queries, different structures used for mobile, and everything else that stems from it. 

The first thing I'd consider is if a `min-width` on the columns (and some programmatic computation around it) would suffice to adapt the desktop view until the mobile (card-based) one sets in. I'd consider passing an optional `isMobile` prop to force the card-based/mobile view on certain viewports (which should cover the Calypso side with the variable sidebar, etc.). On default, the mobile view should set in/out after a certain query change.

For this case (this change), it would help to pass the slug of the plan to highlight given a certain viewport change (and avoid adding more to these media queries). Whatever we create for this will most likely help with what I described above. I think every chance we get right now to do a little bit of refactoring, we should take the opportunity.

^ @southp

aneeshd16 replied:
> So I would wish to see us simplify this to two possible views: mobile, and everything else. 

In that case, we should go for a single flex-based layout. After all, the expanded mobile card is almost exactly the same as the columns on desktop/tablet. We have fixed values for grid width and the number of columns corresponding to the screen width, and those would have to be implemented using media queries. We can continue the discussion in https://github.com/Automattic/wp-calypso/issues/79010, since this is beyond the scope of this PR.

> For this case (this change), it would help to pass the slug of the plan to highlight given a certain viewport change (and avoid adding more to these media queries). 

Maybe I misunderstood this - reacting to a viewport change would still need a media query, right? Can you please explain, with a pseudo-code example, if possible?

chriskmnds replied:
> In that case, we should go for a single flex-based layout. After all, the expanded mobile card is almost exactly the same as the columns on desktop/tablet. We have fixed values for grid width and the number of columns corresponding to the screen width, and those would have to be implemented using media queries. We can continue the discussion in https://github.com/Automattic/wp-calypso/issues/79010, since this is beyond the scope of this PR.

I'm not sure if rewriting the grid implementation is the best use of time to reach the package/NPM milestone. A grid or flex design should have been the consideration from the getgo. Right now, we should be unblocking this while maintaining a decent API over it (at the cost of a bit of extra processing). An internal refactor can happen later if you are up for it. Definitely something for https://github.com/Automattic/wp-calypso/issues/79010. I mentioned the above to give you a bit of context in case it's missing.

> reacting to a viewport change would still need a media query, right

We have viewport / viewport-react packages that you can use. I think the point of investigation would be to determine if sidebar is opened or not. From there, you can add a layout effect and pass the right prop through.

My suggestion is to investigate if possible (and just do it if it is).

aneeshd16 replied:
> I think the point of investigation would be to determine if sidebar is opened or not.

We do have a `getSidebarIsCollapsed` selector that gives us this info.

> From there, you can add a layout effect and pass the right prop through.

An additional challenge here is that in mobile view, the current plan should be at the top of the list. Due the way `plans` and `visiblePlans` are structured and how `planProperties` are calculated, it adds additional complexity to this change. It would all be easier if we move all the plan properties computation to PlanFeaturesMain in this case.

chriskmnds replied:
> It would all be easier if we move all the plan properties computation to PlanFeaturesMain in this case.

These are happening in other PRs already. A couple of hefty refactors will show up soon. I'm not sure it makes a difference though.

For what you are working on, the exploration suggested above would help to also contribute overall, as that's functionality we will need. So avoiding the media query would lead the way for others. 

What is the minimum implementation needed to pass a "spotlight" plan slug (that can be current plan, but also any slug) through from PlansFeaturesMain, on certain viewport, and also have it filtered out of visible plans (also prior to sending through as prop)? Mobile view should be getting none and the full set of visible plans. Is this direction blocked due to something?

chriskmnds reviewed (commented):
I left a few comments to see if we can push this in a better/different direction.

chriskmnds commented on client/my-sites/plan-features-2023-grid/style.scss:
```diff
@@ -416,6 +416,46 @@ $mobile-card-max-width: 440px;
 		}
 	}
 
+	.plan-features-2023-grid__plan-highlight {
```
Let's stop using highlight" and refactor to "spotlight" pls . Corresponding function can be `SpotlightPlan`. We make use of highlight in labels and other places already.

chriskmnds reviewed (changes requested):
It's a block from me. I'm not happy with the implementation here. I'd like to see more perspective on a couple of suggestions left in the comments, particularly on passing through a plan slug (the current-plan slug) to render in a spotlight card, and either sending through visiblePlans without that plan or setting its visibility status to false (as done for plans not visible in tailored flows). I'd also like to see us explore avoid using the media query that will have to be refactored soon (it should be possible with a combination of viewport-react package and knowledge of sidebar status).

I'm also not sure if removing the "spotlight" plan from the plan properties array (that covers the whole/default set of plans) does not create other issues. Assuming I'm reading the [code](https://github.com/Automattic/wp-calypso/pull/78845/files#diff-e268185bf1c246017f6796f5f74ff3c0ae49b12e2acf525d8074859449573a19R1118) correctly.

^ This said, I'm also way over my head on other tasks to keep close tabs on here, as much as I'd like to. If my possibly vague interpretation above is correct, then I'd like to see some changes.

aneeshd16 commented:
Thanks for the feedback @chriskmnds!

> I'd also like to see us explore avoid using the media query that will have to be refactored soon (it should be possible with a combination of viewport-react package and knowledge of sidebar status).

I've tried to address this here: https://github.com/Automattic/wp-calypso/pull/79129

> I'd like to see more perspective on a couple of suggestions left in the comments, particularly on passing through a plan slug (the current-plan slug) to render in a spotlight card, and either sending through visiblePlans without that plan or setting its visibility status to false (as done for plans not visible in tailored flows).

Once we agree on the direction of #79129, I'll update this PR to pass the spotlight slug according to your suggestion.

chriskmnds commented:
Thanks for follow ups @aneeshd16 . This will likely have some conflicts once rebased.

The shape of the plans passed into the grid components hasn't evolved much from current form, but will likely evolve a lot in the next couple of follow-ups. I will try to come up with a structure that will be easy to port over when `plans-features-2023-grid` becomes an index of individual components. 

Currently, we send through the full set of records and a set of "visible" slugs (used to mark the records as "visible" or not). This may change soon. It may be the case we end up with something like `plansForFeaturesGrid`, `plansForComparisonGrid`, `plansForSpotlight`, etc. So you may end up sending a `planForSpotlight` and mark it as hidden (or remove) for `plansForFeaturesGrid`.  I'm currently extracting property computations out of `plans-features-2023-grid`, but still need to conclude on how comparison uses the plan records.

p.s. there's a lot to do [here](https://github.com/Automattic/wp-calypso/pull/79129#issuecomment-1634093632) if you need something to build out more and punt on this 🤷‍♂️

Commit:
Plans: highlight the current plan in the plan features grid

aneeshd16 commented:
Sorry for the mass ping! It tagged all code owners when I changed the base branch.

aneeshd16 commented:
Hi @chriskmnds, based on this feedback(p1689342052022689/1689168387.723419-slack-CV2TX2PAN), I have updated this PR, can you please have a look when you get a chance?

chriskmnds commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -279,14 +286,19 @@ export class PlanFeatures2023Grid extends Component<
+									<div className="plan-features-2023-grid__plan-spotlight">
+										{ this.renderSpotlightPlan( [ spotlightPlanProperties ] ) }
+									</div>
+								) }
```
Let's move this a few layers up please, under `<div className="plans-wrapper">` so it sits on the same level as `plan-features` and `plan-features-2023-grid__plan-comparison-grid-container`. 

`<div className="plan-features">` is what we'll extract to a FeaturesGrid component.

aneeshd16 replied:
Done via [Refactor spotlight plan to be a child of plans-wrapper](https://github.com/Automattic/wp-calypso/pull/78845/commits/f54d0d722c029772e00679ffcca2ece834f3d548)!

chriskmnds commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -450,6 +461,28 @@ export class PlanFeatures2023Grid extends Component<
+	 */
+	renderSpotlightPlan( planPropertiesObj: PlanProperties[] ) {
+		const planCardClasses = classNames(
+			'plan-features-2023-grid__mobile-plan-card',
```
Why is this "mobile" class added to `SpotlightPlan`? Seeing as we are not rendering this in mobile at all, then I suggest creating a separate class to generalise the styles for reuse between mobile table and spotlight plan (if that's the intention).

aneeshd16 replied:
Good catch, it was not required at all. Removed via 3e566e119bd1a69baf76dac3e62bb3b04469aadd.

chriskmnds commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -450,6 +461,28 @@ export class PlanFeatures2023Grid extends Component<
+	 * Similar to `renderMobileView` above.
+	 */
+	renderSpotlightPlan( planPropertiesObj: PlanProperties[] ) {
+		const planCardClasses = classNames(
```
did you mean `spotlightPlanClasses`?

aneeshd16 replied:
Done via https://github.com/Automattic/wp-calypso/pull/78845/commits/94362c7b3d26c012d8df228118677b2df7bf6ba0.

chriskmnds commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -450,6 +461,28 @@ export class PlanFeatures2023Grid extends Component<
+	/**
+	 * Similar to `renderMobileView` above.
+	 */
+	renderSpotlightPlan( planPropertiesObj: PlanProperties[] ) {
```
We only render a single plan inside, so this can be the respective plan's properties alone instead of array.

aneeshd16 replied:
Done via https://github.com/Automattic/wp-calypso/pull/78845/commits/a5c3f0a8b3ed349c8bc8c9b92f26d60a2e4533e2.

chriskmnds commented on client/my-sites/plans-features-main/index.tsx:
```diff
@@ -160,6 +163,13 @@ const OnboardingPricingGrid2023 = ( props: OnboardingPricingGrid2023Props ) => {
 	}
 
+	let spotlightPlanSlug: PlanSlug | undefined;
+	if ( sitePlanSlug && isSpotlightOnCurrentPlan ) {
+		spotlightPlanSlug = Object.keys( planRecords ).find(
+			( planSlug ) => getPlanClass( planSlug ) === getPlanClass( sitePlanSlug )
+		) as PlanSlug | undefined;
+	}
```
Probably missing something, but is double-checking with the plan class necessary? Can't this just be `const spotlightPlanSlug = isSpotlightOnCurrentPlan ? currentPlanSlug : null`?

aneeshd16 replied:
If the current plan is a monthly plan, the spotlight plan can also be the yearly variant of the monthly plan. This is the same conditional used to show the `Upgrade to Yearly` text on the CTAs:
https://github.com/Automattic/wp-calypso/blob/296754eb917d88f12ea27cdbf9a9467648195a09/client/my-sites/plan-features/actions.jsx#L78

chriskmnds replied:
gotcha. thanks for explaining

chriskmnds commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -1038,14 +1072,40 @@ const ConnectedPlanFeatures2023Grid = connect(
+					planPropertiesForMobile[ spotlightPlanIndex ],
+					...planPropertiesForMobile.slice( 0, spotlightPlanIndex ),
+					...planPropertiesForMobile.slice( spotlightPlanIndex + 1 ),
+				];
```
These will change a lot soon, with visibility assignments happening outside of `plan-features-2023-grid`. In fact, features-table will receive only the GridPlan records it needs to render (not the full set, and no `visiblePlans` slugs either). The concept of "planProperties" is also going away..

I wonder if instead of defining these as first-class props on the component, to consider slicing and visibility assignment to happen inside `renderTable` and `renderMobileView`? I think it sounds easier to migrate that way (something that will happen soon).

aneeshd16 replied:
Done via https://github.com/Automattic/wp-calypso/pull/78845/commits/ffa47787b275a7748819061e7fce0a2ee07b7ecf.

worldomonation commented:
Just a side note, I wrote an article on using `git rebase` effectively and Matthew Reishus added an excellent synopsis on why the review bomb happens. p7H4VZ-3Oh-p2#comment-10717

chriskmnds reviewed (changes requested):
Thanks for working on the suggestions. I left a few comments that I'd appreciate if you can take a look please. 

A few details that look a little broken from quick test:

1. highlight labels can overlay logos:

<img width="300" alt="Screenshot 2023-07-17 at 8 32 59 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/32ec1d47-f624-421a-ba70-6b992cdfebaf">

2. how it looks with a WooExpress plan (also sits above the notifications & toggle):

<img width="700" alt="Screenshot 2023-07-18 at 2 08 19 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/4a80b446-6cbf-4df5-870c-29b9bfc70419">

-------------------------

### Design/ Functionality

I also have some functionality concerns I want to raise. Please seek some input in Slack maybe. I'm actually not confident whether we should be shipping anything without these addressed.

1. The features table is meant to be a quick way to compare and visualize plans and their features, including the user's current plan and where it sits in the list/hierarchy e.g. this is information that is essentially lost to me (or hidden behind a price comparison) if we remove the current plan from the grid:

<img width="700" alt="Screenshot 2023-07-18 at 2 15 03 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/bc1604dd-7163-45db-a256-8877097ab7e1">

2. Removing one of the middle plans creates an incomplete view of the features for the immediately higher plan. Not sure if this is something we are aware of e.g. with the Commerce plan, the displayed features are currently an extension of the Business plan's - now removing Business (as current plan) will mention Commerce being an extension to Premium, yet same features shown as previously. There is information that is lost/disconnected AFAIC (and not sure if the solution would be to merge both Business & Commerce features under Commerce - probably not great):

<img width="700" alt="Screenshot 2023-07-18 at 2 21 30 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/1e330eab-7f0e-4138-865b-70e494f3d0dc">

3. Mobile has similar issues and it's not consistent either. For example, current plan is not shown as an "extension" to another plan, yet the features set is the same as previously e.g. Business (current), vs Personal (not current):

| Business (current) | Personal |
| ------------------ | --------- |
| <img width="350" alt="Screenshot 2023-07-18 at 2 40 25 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/9056d174-653c-4c1b-8342-529a24cf7716"> | <img width="350" alt="Screenshot 2023-07-18 at 2 40 14 PM" src="https://github.com/Automattic/wp-calypso/assets/1705499/ff8b6740-85f6-44e0-b043-787412e84843"> |

saygunnyc commented:
Thank you for the notes @chriskmnds, I'll try to elaborate on each below.

> The features table is meant to be a quick way to compare and visualize plans and their features, including the user's current plan and where it sits in the list/hierarchy e.g. this is information that is essentially lost to me (or hidden behind a price comparison) if we remove the current plan from the grid:

Once the user is on the Plans page within Calypso, we'd like the user to easily separate their plan from the rest of the options. It becomes less obvious with several tags, recommended plans, and such. We also would like to separate it from the options you can pick moving forward. This plan card above also evolves and can keep evolving into carrying more options within the plan you selected, like storage. Comparison for sure is a feature we'd like to provide that can be accomplished with the CTA below, "`Compare my plan`" Also noting this is something we've discussed with @pablohoneyhoney a while ago, and agreed upon.


> Removing one of the middle plans creates an incomplete view of the features for the immediately higher plan. Not sure if this is something we are aware of e.g. with the Commerce plan, the displayed features are currently an extension of the Business plan's - now removing Business (as current plan) will mention Commerce being an extension to Premium, yet same features shown as previously. There is information that is lost/disconnected AFAIC (and not sure if the solution would be to merge both Business & Commerce features under Commerce - probably not great):

We're not removing a plan, we're only separating user's plan from their options to select from. The user has already selected that plan `Premium` in this case and is benefiting from the features the plan offers, so I don't see this as an issue. They can still look at the upper-tier features indicated with `"Everything in Premium, plus" ` There's nothing wrong referring to user's existing plan, without breaking it down fully. And they can always compare their plan by clicking on the CTA below if they need more help or information or a more detailed comparison.


> Mobile has similar issues and it's not consistent either. For example, current plan is not shown as an "extension" to another plan, yet the features set is the same as previously e.g. Business (current), vs Personal (not current):

The mobile view, something we can/should further iterate moving forward, is not a comparison layout. If users would like to compare plans, there's already a separate view, as mentioned above. While it is OK not to have 💯 parity between mobile and desktop viewports, we could remove the user's plan features from mobile if we'd like to have full parity. (Likely unnecessary, as we could learn from user behavior on each) We have another tab called My Plan on Plans that we never have gotten to re-designing yet. That might be a better place to showcase the user's existing plan's features while guiding towards upsells, etc. We should get to it sooner. And it might be a good idea to link to the mentioned plan features if the user clicks on "Everything on Free, plus.." linking to Free plan features.

Commit:
Hide header logo

Commit:
Refactor spotlight plan to be a child of plans-wrapper

Commit:
This patch removes the CSS class "plan-features-2023-grid__mobile-plan-card" from the PlanFeatures2023Grid component, as it is not needed for the spotlight plan rendering.

Commit:
Rename variable for spotlight plan card class

Commit:
Change props to object instead of array

Commit:
Refactor, move planProperties slicing to renderXYZ functions

aneeshd16 commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -350,38 +358,34 @@ export class PlanFeatures2023Grid extends Component<
+		);
 
 		const numberOfPlansToShowOnTop = 4 === plansToShow.length ? 2 : 3;
-		const topRowPlans = plansToShow.slice( 0, numberOfPlansToShowOnTop );
-		const bottomRowPlans = plansToShow.slice( numberOfPlansToShowOnTop, plansToShow.length );
-		const planPropertiesForTopRow = planProperties.filter( ( properties: PlanProperties ) =>
-			topRowPlans.includes( properties.planName )
-		);
-		const planPropertiesForBottomRow = planProperties.filter( ( properties: PlanProperties ) =>
-			bottomRowPlans.includes( properties.planName )
-		);
```
This change was not related to the current PR, but it simplifies this part.

aneeshd16 commented:
Thanks for the detailed feedback, @chriskmnds! Much appreciated! I have addressed the code feedback and linked to the individual commits in your comments.

> highlight labels can overlay logos:

Good catch! The logos are supposed to be hidden on the spotlight plan. Here's the design ref: dyXNnR0oHu8H47eUES8ktJ-fi-362%3A19122. I have fixed this via https://github.com/Automattic/wp-calypso/pull/78845/commits/4d4fb1b2b35e5399a9805d9e23529b31eafbf23e.

> how it looks with a WooExpress plan (also sits above the notifications & toggle):

I was confused as well, but this is from a different component and not rendered by the spotlight component added in this PR. Here's a production screenshot:
<img width="1916" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/ff498284-33e0-42ab-9ec7-337f0fd303fb">

chriskmnds commented:
> Once the user is on the Plans page within Calypso, we'd like the user to easily separate their plan from the rest of the options. It becomes less obvious with several tags, recommended plans, and such. We also would like to separate it from the options you can pick moving forward. This plan card above also evolves and can keep evolving into carrying more options within the plan you selected, like storage. Comparison for sure is a feature we'd like to provide that can be accomplished with the CTA below, "Compare my plan" Also noting this is something we've discussed with @pablohoneyhoney a while ago, and agreed upon.

Thanks. I think my point was simple, that we lose communicating that "previous tier" / "next tier" info (from the user’s current plan) by removing current plan from the table. It’s not visible in the comparison grid either (which reorders plans and places current one first to the left). 

p.s. sounds like something WooExpress plans page also doing differently (keeping the current plan in the table).

> There's nothing wrong referring to user's existing plan, without breaking it down fully. 

@saygunnyc I think you’ve misread my comment (user is not on Premium, but on Business I the linked image). In any case, if we go by what you mention  then it sounds like a technical oversight — Commerce changed from "Everything in Business, plus" to "Everything in Premium, plus" because of current plan being Business. is that fine without updating feature set for Commerce?

> I was confused as well, but this is from a different component and not rendered by the spotlight component added in this PR. Here's a production screenshot:

My impression is we should have consistent layout in plans page irrespective of visiting with a Woo site or Newsletter or standard one. They all use same components now, they just got to this update first (spotlight for current plan). I don’t see reason to maintain separate components for same purpose, especially if there’s no real intent behind. 

Another thing to consider about the Woo plans is whether to also remove the current plan from the table (assuming that’s something we will continue doing for everything else).

chriskmnds reviewed (commented):
Great job on the code @aneeshd16!

I think there may be a couple of things to address or follow up on from my previous comment/response.

saygunnyc commented:
> wehther we lose communicating that "previous tier" / "next tier" info (from the user’s current plan) by removing current plan from the table. It’s not visible in the comparison grid either (which reorders plans and places current one first to the left).

I think I get what you mean, but I don't think we lose it as we don't say previous or next; we mention the plan name when picking a baseline for elaborating on the plusses. Perhaps we can say Everything on Xplan + Yplan as we move Y plan above if you think it helps.  The comparison in this context should always put the user's plan into the left column, comparing it with other plans as the default behavior. 


>— Commerce changed from "Everything in Business, plus" to "Everything in Premium, 

This should not be the case because now the order is different. It should always keep informing the right thing. Commerce should always be Everything in Business, plus. no matter if we moved Business above. Or as suggested above maybe we then use both plan names.

We can of course, find better ways in both separating the plan out and listing out features in a better way. I actually shared a P2 yesterday regarding further iterations and experiments on the plans grid. p9Jlb4-7RP-p2


From what I see from the screenshot from Woo,
It's weird to me that they repeat the plan both in the grid and above. It also is weird that one is labeled as `My Plan` while the other one is `Your plan`They also have different plan descriptions on above and below.


> My impression is we should have consistent layout in plans page irrespective of visiting with a Woo site or Newsletter or standard one. 

Couldn't agree more, especially as it's possible for a user to move from one to another.

Commit:
Refactor PlanFeatures2023Grid component to pass previousProductName as a prop to each plan card.

aneeshd16 commented:
I have fixed the `Everything in <plan name>` text so that the previous plan name always refers to the correct plan, even if the previous plan has been removed from the table. Done via https://github.com/Automattic/wp-calypso/pull/78845/commits/0cda1d15a0a450db952e165582b4277abb0682fe.

@chriskmnds and @saygunnyc, can you please test and let me know if this is okay? Thanks!

**Before**
<img width="974" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/93143543-94b2-4c00-b7bd-562ce9293fbe">

**After**
<img width="953" alt="image" src="https://github.com/Automattic/wp-calypso/assets/5436027/f7287a79-99d2-4e48-9e8e-bfac99a94820">

aneeshd16 commented:
> From what I see from the screenshot from Woo,
It's weird to me that they repeat the plan both in the grid and above. It also is weird that one is labeled as My Plan while the other one is Your planThey also have different plan descriptions on above and below.

> My impression is we should have consistent layout in plans page irrespective of visiting with a Woo site or Newsletter or standard one.

Thanks for highlighting this @chriskmnds and @saygunnyc! I have created a follow-up issue to discuss this further here: https://github.com/Automattic/martech/issues/2001

chriskmnds committed:
retrieve plan index for previousProductName from original plan properties

chriskmnds commented on client/my-sites/plan-features-2023-grid/index.tsx:
```diff
@@ -1036,16 +1070,31 @@ const ConnectedPlanFeatures2023Grid = connect(
+					};
+				}
+				return properties;
+			} );
```
We shouldn't need to do this at the data structure level (adding another property to refer to the previous record in the list). It creates a coupling between the records, which are to be indexed at the Set/array level.

We can retrieve the index from the properties array and skip extending the data structure. I pushed a commit to do this, so can you please test the changes to confirm no breakage?

p.s. There is technically/conceptually coupling obviously, either way, so your approach isn't wrong in any way (since plan features extend from plan to plan). It's just to help with migrating this to the refactors i  the other PR that moves properties to GridPlan concept. We can explore something like "extendsFrom" maybe in the features structure there, which would save us the need to build it out of the order 🤔

chriskmnds reviewed (approved):
LGTM. Thanks for the follow-up ticket! stellar job on the follow-ups and the explorations in other PR. thanks for bearing with me! 👍 ⚡

I pushed a commit to address [comment](https://github.com/Automattic/wp-calypso/pull/78845#discussion_r1276326675) and align things better with how we perceive these data structures to evolve. Let's give this another round of testing and ship if everything ok. I tested and looks ok, but might have missed something.
